assets/js/googlesitekit/widgets/datastore/areas.js

```diff
@@ -0,0 +1,215 @@
+		invariant( areaSlug, 'areaSlug is required.' );
+		invariant( settings, 'settings is required.' );
+		invariant( settings.title, 'settings.title is required.' );
+		invariant( settings.subtitle, 'settings.subtitle is required.' );
+		invariant( settings.style, 'settings.style is required.' );
+
+		const settingsToUse = {
+			priority: 10,
+			...settings,
+		};
+
+		return {
+			payload: { areaSlug, settings: settingsToUse },
+			type: REGISTER_WIDGET_AREA,
+		};
+	},
+};
+
+export const controls = {};
+
+export const reducer = ( state, { type, payload } ) => {
+	switch ( type ) {
+		case ASSIGN_WIDGET_AREA: {
+			const { areaSlug, contextSlugs } = payload;
+
+			const { contexts } = state;
+			contextSlugs.forEach( ( contextSlug ) => {
+				if ( contexts[ contextSlug ] === undefined ) {
+					return;
+				}
+
+				if ( ! contexts[ contextSlug ].includes( areaSlug ) ) {
+					contexts[ contextSlug ].push( areaSlug );
+				}
+			} );
+
+			return {
+				...state,
+				contexts,
+			};
+		}
+
+		case REGISTER_WIDGET_AREA: {
+			const { areaSlug, settings } = payload;
+
+			if ( state.areas[ areaSlug ] !== undefined ) {
+				global.console.warn( `Could not register widget area with slug "${ areaSlug }". Widget Area "${ areaSlug }" is already registered.` );
+
+				return { ...state };
+			}
+
+			return {
+				...state,
+				areas: {
+					...state.areas,
+					[ areaSlug ]: { ...settings, slug: areaSlug },
+				},
+			};
+		}
+
+		default: {
+			return { ...state };
+		}
+	}
+};
+
+export const resolvers = {
+};
+
+export const selectors = {
+	/**
+	 * Checks if a widget area has been registered.
+	 *
+	 * Returns `true` if the widget area has been registered.
+	 * Returns `false` if the widget area has NOT been registered.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {Object} state    Data store's state.
+	 * @param  {string} areaSlug Widget Area's slug.
+	 * @return {boolean}         `true`/`false` based on whether widget area has been registered.
+	 */
+	isWidgetAreaRegistered( state, areaSlug ) {
+		const { areas } = state;
+
+		return areas[ areaSlug ] !== undefined;
+	},
+
+	/**
+	 * Returns all widget areas registered for a given context.
+	 *
+	 * Returns an array of all widget areas for a given context.
+	 * The widget areas are returned in order of their priority, so can be rendered in
+	 * the order provided by the selector.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param  {Object} state       Data store's state.
+	 * @param  {string} contextSlug Widget context to get areas for.
+	 * @return {Array}              An ordered array of widget areas for this context.
+	 */
+	getWidgetAreas( state, contextSlug ) {
+		invariant( contextSlug, 'contextSlug is required.' );
+
+		const { areas, contexts } = state;
+
+		return Object.values( areas ).filter( ( area ) => {
+			return contexts[ contextSlug ] && contexts[ contextSlug ].includes( area.slug );
+		} ).sort( ( areaA, areaB ) => {
+			return ( areaA.priority >= areaB.priority ) ? 1 : -1;
+		} );
```

Comment:
It may be faster to do the sorting whenever a new widget area is registered? The selector will be called over and over again, where registration is a more distinct action. We'll register several widget areas so we'll still get multiple sort calls, but overall I feel doing that in the action reducer is more appropriate.

Reply:
Ultimately we'd be best-served with something like `reselect`, to memoise our selector calls. We can order the object by key, but I'd prefer the logic of how to select the data is stored in the selector, not in the data store. That's mixing the raw storage of the data with how a _specific_ selector should present it, which isn't how one should use Redux.

We're better off looking into optimising our selector calls in the future than storing things sorted in the datastore, because that breaks down once we want to sort things another way (unlikely in this case, but still 🙂).